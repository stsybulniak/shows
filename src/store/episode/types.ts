export enum EpisodeTypes {
  fetchEpisodeList = '@@episode/FETCH_EPISODE_LIST',
  fetchEpisodeListSuccess = '@@episode/FETCH_EPISODE_LIST_SUCCESS',
  fetchEpisodeListFailure = '@@episode/FETCH_EPISODE_LIST_FAILURE',
}

export interface IEpisode {
  id: number;
  name: string;
  image: { medium: string; original: string } | null;
  summary: string;
  number: number;
  season: number;
}

export interface IEpisodeState {
  episodes: { [key: number]: IEpisode };
  isLoading: boolean;
  error: string | null;
}
