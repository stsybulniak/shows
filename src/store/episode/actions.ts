import { action } from 'typesafe-actions';
import { EpisodeTypes, IEpisode } from './types';

export const fetchEpisodeList = () => action(EpisodeTypes.fetchEpisodeList);
export const fetchEpisodeListSuccess = (payload: { [key: string]: IEpisode }) =>
  action(EpisodeTypes.fetchEpisodeListSuccess, payload);
export const fetchEpisodeListFailure = (payload: string) => action(EpisodeTypes.fetchEpisodeListFailure, payload);
