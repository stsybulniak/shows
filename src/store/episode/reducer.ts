import { Reducer } from 'redux';
import { IEpisodeState, EpisodeTypes } from './types';

export const initialState: IEpisodeState = {
  isLoading: false,
  episodes: {},
  error: null,
};

export const episodeReducer: Reducer<IEpisodeState> = (state = initialState, { type, payload = {} }) => {
  switch (type) {
    case EpisodeTypes.fetchEpisodeList: {
      return { ...state, isLoading: true };
    }
    case EpisodeTypes.fetchEpisodeListSuccess: {
      return { ...state, isLoading: false, episodes: payload };
    }
    case EpisodeTypes.fetchEpisodeListFailure: {
      return { ...state, isLoading: false, episodes: {}, error: payload };
    }
    default: {
      return state;
    }
  }
};
