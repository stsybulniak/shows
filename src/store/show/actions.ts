import { action } from 'typesafe-actions';
import { IShow, ShowTypes } from './types';

export const fetchShow = () => action(ShowTypes.fetchShow);
export const fetchShowSuccess = (payload: IShow) => action(ShowTypes.fetchShowSuccess, payload);
export const fetchShowError = (payload: string) => action(ShowTypes.fetchShowFailure, payload);
