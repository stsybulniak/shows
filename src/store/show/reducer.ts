import { Reducer } from 'redux';
import { IShowState, ShowTypes } from './types';

export const initialState: IShowState = {
  isLoading: false,
  show: null,
  error: null,
};

export const showReducer: Reducer<IShowState> = (state = initialState, { type, payload = {} }) => {
  switch (type) {
    case ShowTypes.fetchShow: {
      return { ...state, isLoading: true };
    }
    case ShowTypes.fetchShowSuccess: {
      return { ...state, isLoading: false, show: payload };
    }
    case ShowTypes.fetchShowFailure: {
      return { ...state, isLoading: false, show: null, error: payload };
    }
    default: {
      return state;
    }
  }
};
