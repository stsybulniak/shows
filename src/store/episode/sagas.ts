import { all, call, fork, put, takeEvery } from 'redux-saga/effects';
import { IEpisode, EpisodeTypes } from './types';
import axios, { AxiosResponse } from 'axios';
import { fetchEpisodeListFailure, fetchEpisodeListSuccess } from './actions';
import { convertEpisodeToMap } from '../../utils';

const fetchEpisodes = () => axios.get(`${process.env.REACT_APP_API_URL}/episodes`)

export function* handleFetchEpisodeList() {
  try {
    const resp: AxiosResponse<IEpisode[]> = yield call(fetchEpisodes);
    const data = convertEpisodeToMap(resp.data);
    yield put(fetchEpisodeListSuccess(data));
  } catch (err) {
    yield put(fetchEpisodeListFailure(err.message));
  }
}

function* watchEpisodeSaga() {
  yield takeEvery(EpisodeTypes.fetchEpisodeList, handleFetchEpisodeList);
}

function* episodeSaga() {
  yield all([fork(watchEpisodeSaga)]);
}

export default episodeSaga;
