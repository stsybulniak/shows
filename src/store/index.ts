import { combineReducers } from 'redux';
import { all, fork } from 'redux-saga/effects';
import { connectRouter, RouterState } from 'connected-react-router';
import { History } from 'history';
import showSaga from './show/sagas';
import { IShowState, showReducer } from './show';
import episodeSaga from './episode/sagas';
import { episodeReducer, IEpisodeState } from './episode';

export interface IApplicationState {
  router: RouterState;
  show: IShowState;
  episode: IEpisodeState;
}

export const createRootReducer = (history: History) =>
  combineReducers({
    router: connectRouter(history),
    show: showReducer,
    episode: episodeReducer,
  });

export function* rootSaga() {
  yield all([fork(showSaga), fork(episodeSaga)]);
}
