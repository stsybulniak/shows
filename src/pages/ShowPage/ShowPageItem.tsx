import { FC, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ErrorOrLoading from '../../components/ErrorOrLoading/ErrorOrLoading';
import Show from '../../components/Show/Show';
import { IApplicationState } from '../../store';
import * as showActions from '../../store/show/actions';
import { viewportWidth } from '../../utils';

const ShowPage: FC = () => {
  const { show, isLoading, error } = useSelector((state: IApplicationState) => state.show);

  const dispatch = useDispatch();

  useEffect(() => {
    if (!show) {
      dispatch(showActions.fetchShow());
    }
  }, [show]);

  return (
    <ErrorOrLoading isLoading={isLoading} error={error}>
      {show && <Show show={show} viewportWidth={viewportWidth} />}
    </ErrorOrLoading>
  );
};

export default ShowPage;
