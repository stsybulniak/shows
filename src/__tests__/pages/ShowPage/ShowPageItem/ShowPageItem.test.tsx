import { render } from '@testing-library/react';
import ShowPageItem from '../../../../pages/ShowPage/ShowPageItem';
import * as reactRedux from 'react-redux';
import { ShowTypes } from '../../../../store/show';
import { EpisodeTypes } from '../../../../store/episode';

const showMock = {
  id: 1,
  name: 'Test show name',
  image: {
    medium: 'https://static.tvmaze.com/uploads/images/medium_landscape/115/288994.jpg',
    original: 'https://static.tvmaze.com/uploads/images/original_untouched/115/288994.jpg',
  },
  summary: 'Some description',
};

describe('<ShowPageItem />', () => {
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
  });

  it('Render initial show item page', () => {
    const dispatchMock = jest.fn();
    useSelectorMock.mockReturnValue({ show: null, isLoading: true, error: null });
    useDispatchMock.mockReturnValue(dispatchMock);
    render(<ShowPageItem />);
    expect(dispatchMock).toHaveBeenCalledWith(expect.objectContaining({ type: ShowTypes.fetchShow }));
  });

  it('Render show item page with sucsessfull show fetch', () => {
    const dispatchMock = jest.fn();
    useSelectorMock.mockReturnValue({ show: showMock, isLoading: false, error: null, episodes: {} });
    useDispatchMock.mockReturnValue(dispatchMock);
    render(<ShowPageItem />);
    expect(dispatchMock).toHaveBeenCalledWith(expect.objectContaining({ type: EpisodeTypes.fetchEpisodeList }));
  });

  it('Render show item page with faild show fetch', () => {
    const dispatchMock = jest.fn();
    useSelectorMock.mockReturnValue({ show: null, isLoading: false, error: 'Error here', episodes: {} });
    useDispatchMock.mockReturnValue(dispatchMock);
    const { baseElement } = render(<ShowPageItem />);
    expect(baseElement).toMatchSnapshot();
  });
});
