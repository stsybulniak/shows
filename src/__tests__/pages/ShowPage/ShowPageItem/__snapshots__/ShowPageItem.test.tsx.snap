// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShowPageItem /> Render show item page with faild show fetch 1`] = `
<body>
  <div>
    <h3>
      Error here
    </h3>
  </div>
</body>
`;
